onnxruntime/test/shared_lib/test_model_builder_api.cc

```diff
@@ -0,0 +1,485 @@
+
+    // create ValueInfo and release the type info as CreateValueInfo takes a copy.
+    OrtValueInfo* input_value_info = nullptr;
+    Ort::ThrowOnError(model_builder_api.CreateValueInfo("X", input_type_info, &input_value_info));
+    api.ReleaseTypeInfo(input_type_info);  // input_value_info took a copy
+    tensor_type_info = nullptr;
+
+    // model outputs
+    OrtTypeInfo* output_type_info = nullptr;
+    std::vector<int64_t> output_dims = {3, 3};
+
+    Ort::ThrowOnError(api.CreateTensorTypeAndShapeInfo(&tensor_type_info));
+    Ort::ThrowOnError(api.SetTensorElementType(tensor_type_info, ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT));
+    Ort::ThrowOnError(api.SetDimensions(tensor_type_info, output_dims.data(), output_dims.size()));
+
+    Ort::ThrowOnError(api.CreateTensorTypeInfo(tensor_type_info, &output_type_info));
+    api.ReleaseTensorTypeAndShapeInfo(tensor_type_info);  // input_type_info took a copy
+
+    OrtValueInfo* output_value_info = nullptr;
+    Ort::ThrowOnError(model_builder_api.CreateValueInfo("Z", output_type_info, &output_value_info));
+    api.ReleaseTypeInfo(output_type_info);
+
+    std::vector<OrtValueInfo*> graph_inputs = {input_value_info};
+    std::vector<OrtValueInfo*> graph_outputs = {output_value_info};
+    Ort::ThrowOnError(model_builder_api.SetGraphInputs(graph, graph_inputs.data(), graph_inputs.size()));
+    Ort::ThrowOnError(model_builder_api.SetGraphOutputs(graph, graph_outputs.data(), graph_outputs.size()));
+    input_value_info = nullptr;  // graph now owns the input/output values
+    output_value_info = nullptr;
+
+    //
+    // Gemm node
+    //
+
+    OrtOpAttr* alpha_attr = nullptr;
+    float alpha_value = 2.0;
+    Ort::ThrowOnError(api.CreateOpAttr("alpha", &alpha_value, 1, OrtOpAttrType::ORT_OP_ATTR_FLOAT, &alpha_attr));
+
+    std::vector<const char*> node_input_names = {"X", "Y"};
+    const std::string gemm_output_name = use_constant_node ? "Z_temp" : "Z";
+    std::vector<const char*> node_output_names = {gemm_output_name.c_str()};
+    std::vector<OrtOpAttr*> node_attributes{alpha_attr};
+    OrtNode* node = CreateNode(model_builder_api, "Gemm", "Gemm1", node_input_names, node_output_names, node_attributes);
+
+    api.ReleaseOpAttr(alpha_attr);  // CreateNode copies all OrtOpAttr instances
+
+    Ort::ThrowOnError(model_builder_api.AddNodeToGraph(graph, node));
+    node = nullptr;  // graph now owns node
+
+    // Y input
+    std::vector<int64_t> y_dims = {2, 3};
+    deleter.weights.emplace_back(
+        std::make_unique<std::vector<float>>(std::initializer_list<float>{1.0f, 2.0f, 3.0f,
+                                                                          4.0f, 5.0f, 6.0f}));
+    auto& y_values = *deleter.weights.back();
+
+    // create an initializer for the Y input. add to `weights` so the memory remains valid
+    OrtValue* y_tensor = nullptr;
+    auto info = Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeDefault);
```

Comment:
`MemoryInfo` seems not to be used when calling `CreateTensorWithDataAndDeleterAsOrtValue()`?

Reply:
Memory info comes from the OrtAllocator in the deleter.

Reply:
Then this `info` variable is unnecessary?

Reply:
Ah no. This code initially used CreateTensorWithDataAsOrtValue but I changed it to CreateTensorWithDataAndDeleterAsOrtValue to test that functionality and overlooked removing the `info` variable. Will remove.